[fx] fix merge_matmul tests making invalid torch.split calls

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74728
* #74727
* __->__ #75019
* #76675

The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

peterbell10 committed:
[fx] fix merge_matmul tests making invalid torch.split calls

The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Rebase and move TLS on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Fix merge conflicts on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Rebase and fix merge conflicts on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Rebase and fix merge conflicts on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

jamesr66a reviewed (approved):
Original author of this pass doesn't work here anymore but I'm rubber stamping

maybe cc @wushirong @jfix71 want to take a look as well

peterbell10 committed:
Rebase and fix merge conflicts on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

jfix71 commented on torch/fx/experimental/merge_matmul.py:
```diff
@@ -9,20 +9,27 @@
+        List of matmul results for each input tensor.
     """
-    return t.shape[0]
+    splits = [x.shape[0] for x in inputs]
+
+    # When fx tracer is running, x.shape[0] will be torch.fx.Attribute but we
+    # need an int even when tracing
+    if isinstance(result, torch.fx.Proxy):
+        splits = [0] * len(inputs)
```
Tiny nit, marginally inefficient when tracing 🙂 
```suggestion
    # When fx tracer is running, x.shape[0] will be torch.fx.Attribute but we
    # need an int even when tracing
    if isinstance(result, torch.fx.Proxy):
        splits = [0] * len(inputs)
    else:
        splits = [x.shape[0] for x in inputs]
```

jfix71 reviewed (approved):
LGTM!

peterbell10 committed:
Update on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[fx] fix merge_matmul tests making invalid torch.split calls"


The merge_matmul tests currently call `torch.split` with a list of
`torch.fx.Attribute` objects which won't work once `torch.split` is moved into
C++ because the argument parser actually checks the list contains integers
before calling `__torch_function__`.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge this please
